wallet.go

```diff
@@ -0,0 +1,138 @@
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"golang.org/x/crypto/sha3"
+)
+
+type ChannelState uint8
+
+const (
+	ChannelStateCooperative ChannelState = iota
+	ChannelStateChallenge
+	ChannelStateIndependent
+)
+
+type Participant uint8
+
+const (
+	ParticipantOwner Participant = iota
+	ParticipantIntermediary
+)
+
+const CurrentBlockNumber = 0 // todo: replace with a service
+
+var myAddress common.Address = common.Address{} // todo: compute from injected private key - cli? config file?
+
+// ChannelWallet is the off-chain accounting struct for a state channel wallet.
+type ChannelWallet struct {
+	//////////////////
+	// Channel Info //
+	//////////////////
+
+	ChannelAddress common.Address
+
+	Owner        common.Address
+	Intermediary common.Address
+
+	////////////////////
+	// Dynamic State  //
+	////////////////////
+
+	State      ChannelState
+	TurnNumber uint64
+
+	///////////////////
+	// Balances Info //
+	///////////////////
+
+	// IntermediaryBalance is the wallet balance that the Intermediary has rights to.
+	// The wallet owner implicitly has claim over the rest.
+	//
+	// This representation is limited to a single-asset channel.
+	IntermediaryBalance *big.Int
+
+	// A list of "running" HTLCs. i.e. HTLCs that have been initiated but
+	// not yet claimed or expired.
+	HTLCs []HTLC
+}
+
+func (cw *ChannelWallet) IncrementTurnNumber() {
+	cw.TurnNumber++
+
+	// todo: abi-encode the state, sign it, and communicate it to the other party
+}
+
+func (cw *ChannelWallet) ClearExpiredHTLCs() {
+	retainedHTLCs := make([]HTLC, 0, len(cw.HTLCs))
+
+	for _, htlc := range cw.HTLCs {
+		// NOT expired - keep it
+		if CurrentBlockNumber < htlc.ExpirationBlock {
+			retainedHTLCs = append(retainedHTLCs, htlc)
+		}
+	}
+
+	if len(retainedHTLCs) != len(cw.HTLCs) {
+		cw.HTLCs = retainedHTLCs
+		cw.IncrementTurnNumber()
+	}
+}
+
+func (cw *ChannelWallet) ReleaseHTLC(preimage []byte) {
+	// Check for a LN compatible preimage
+	sha := sha3.New256()
+	sha.Write(preimage)
+	lnHash := common.Hash(sha.Sum(nil))
+
+	// Check for an EVM compatible preimage
+	kec := sha3.NewLegacyKeccak256()
+	kec.Write(preimage)
+	evmHash := common.Hash(kec.Sum(nil))
+
+	for i, htlc := range cw.HTLCs {
+		if htlc.Hash == lnHash || htlc.Hash == evmHash {
+			if CurrentBlockNumber >= htlc.ExpirationBlock {
+				// Expired - do not honor. Log
+				continue
+			}
+
+			// increment the intermediary's claimable balance if the intermediary is the recipient.
+			//
+			// If the owner is the recipient, the owner regains implicit control over the funds
+			// previously locked by the HTLC.
+			if htlc.ReleaseTo == ParticipantIntermediary {
+				cw.IntermediaryBalance.Add(cw.IntermediaryBalance, htlc.Amount)
+			}
+
+			// remove the running HTLC
+			cw.HTLCs = append(cw.HTLCs[:i], cw.HTLCs[i+1:]...)
+
+			cw.IncrementTurnNumber()
+		}
+	}
+}
+
+// SignedChannelState is a ChannelState with a signature from the owner.
+type SignedChannelState struct {
```

Comment:
This is intended as a wire format - eg, communicate this struct with your channel counterparty off chain.

The adding or removal of HTLCs is encoded within the struct. If I locally add an HTLC, I add it to the state, sign, and send to my counterparty (rather than signing an individual HTLC).